Adding support for * as destinations

Hanstad opened the pull request:
Distributing build to all distribution groups by using the '*' as destination. Make exceptions by adding distributions groups to destination_excludes

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/fastlane-plugin-appcenter?pullRequest=150) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/fastlane-plugin-appcenter?pullRequest=150) before we can accept your contribution.<br/><br/>:x: Hanstad [sign now](https://cla.opensource.microsoft.com/microsoft/fastlane-plugin-appcenter?pullRequest=150)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/fastlane-plugin-appcenter?pullRequest=150) it.</sub>

phillipleblanc commented:
This isn't very clear that the '*' only applies to distribution groups. We also support distributing to Store destinations and testers directly.

Hanstad commented:
> This isn't very clear that the '*' only applies to distribution groups. We also support distributing to Store destinations and testers directly.

What do you suggest that we do? Will it be possible to support Store as well? Or should we introduce yet another option?

Hanstad commented:
@b3hni4

phillipleblanc commented on lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb:
```diff
@@ -596,7 +620,17 @@ def self.example_code
+            destinations: "*",
+            destination_excludes: "Collaborators",
+            destination_type: "store",
+            release_notes: "this is a store release"
```
This example is wrong - it would be incorrect to have a destination as "*" for a store release

Hanstad replied:
I agree. I need to change the destination_type

jpeg-ms replied:
And also `.aab` is only valid for google play :)
The example Fastfile should also be modified to show off this feature.

phillipleblanc commented:
> > This isn't very clear that the '*' only applies to distribution groups. We also support distributing to Store destinations and testers directly.
> 
> What do you suggest that we do? Will it be possible to support Store as well? Or should we introduce yet another option?

I can't imagine a scenario in which anyone would want to distribute to all Store destinations, since that includes Production as well as any beta testing groups supported by the Stores.

Hanstad commented:
I could change the description of the option to state that this only works for distribution groups. How does that sound? If it does not make sense for Store lists, but does make sense for Distrubution lists, then that is the only option unless you will go splitting the option into a distribution-list and a store-list and by that removing the backwards compatibility

phillipleblanc commented:
Another way would be to error out if `*` was specified for the destinations and the `destination_type` wasn't `group`

jpeg-ms commented on lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb:
```diff
@@ -191,7 +192,24 @@ def self.run_release_upload(params)
 
-            destinations_array = destinations.split(',')
+            destinations_array = []
+            if destinations == '*'
```
At this point we must make sure that the distribution type is `group`.

Hanstad replied:
Will add this :)

Ajaykn21 commented on lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb:
```diff
@@ -108,6 +108,7 @@ def self.run_release_upload(params)
         owner_type = params[:owner_type]
         app_name = params[:app_name]
         destinations = params[:destinations]
+        destination_excludes = params[:destination_excludes]
```
Hi @Hanstad 
Agree with @jpeg-ms, when we can achieve it with existing options itself.
Please check if you can remove `destination_excludes` .

Hanstad commented:
I have implemented an error-out if destinations are set to '*' while destination_type is not 'group'.
Furthermore I have implemented a Important message if the destination_excludes contains groups that are not found on AppCenter.

Hanstad commented:
I would appreciate if someone could give feedback or merge my PR

jpeg-ms commented on lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb:
```diff
@@ -191,7 +193,29 @@ def self.run_release_upload(params)
+                UI.important("Could not find the excluded distribution group: #{group}") unless distribution_group_names.include? group.downcase 
+              end
+            else
+              destinations_array = destinations.split(',')
```
It might be worth it adding a `map(&:strip)` here as well while we're at it.

Hanstad replied:
map(&:strip) added as well as unit tests

Hanstad commented:
Thanks for your questions. 

Use case: 
We have a app that is build using fastlane on different branches (development, rc, master etc). Each build is uploaded to is own app on AppCenter (app_master, app_development etc.) but uses the same Fastfile.
Depending on the app on AppCenter (app_master, app_development etc.) different groups of people receives the build. This list changes all the time. 
We would like to control who gets the build on AppCenter by adding or removing distribution groups and not change the fastlane code everytime. 

I'm not a fan of the exclude either, but the list received from `fetch_distribution_groups` contains the **Collaborators** group, which I don't want to distribute to. This is a group that I cannot remove (AFAIK). So I introduced the exclude to remove this group. An alternative would be to hardcode the removal of **Collaborators**

Please let me know which approach you prefer.

I will start working on unit tests for this.

jpeg-ms commented:
Interesting, thanks the detailed response.
It sounds to me like you could use the environment variable `APPCENTER_DISTRIBUTE_DESTINATIONS` instead of modifying the Fastfile, to select which groups to send the release to.

Again out of curiosity, why not distribute to Collaborators? As owners of the app, shouldn't they get notified when there's a new version?

Hanstad commented:
Thanks for your proposal. The number of groups changes all the time, so if I used the env. variable I would have to update the Jenkinsfile (we are using Jenkins) all the time (equal to update the Fastfile as it will require a commit) or update all our build agents all the time. So I don't think that is a good solution.

Collaborators are all people with interest of the app but not all of these needs the build. Some looks at diagnostics and some looks at analytics only.

jpeg-ms commented:
Hey @hanstad

I'm sorry to say this but we aren't too excited about the excludes part. Adding support for `'*'` seems OK to us, but excludes seems like a non-necessary feature. It will not be clear how that works with standard `destinations`, and that doesn't seem like a feature we want to have to maintain in this plugin.
Note that in your case, Collaborators all have access to the app and all releases anyway, so it's not a security feature. Users who don't want to receive notifications about new releases can opt out from email notifications themselves.

Alternatively you could look into using the CLI to list the groups prior to calling fastlane, or even distribute using the CLI.

Hanstad commented:
Hey @jpeg-ms 
I'm sad about your decision, but that is sometimes how it is when using 3'rd party plugins. I will update the PR to exclude the exclude option. Are there anyway that I can introduce a "secret" env variable for the excludes instead?

jpeg-ms commented:
If we introduce something that isn't documented (secret as you say), then we will end up not maintaining it, which would only make it worse for all of us on the longer term.

Hanstad commented:
I have now removed the support for destination_excludes

jpeg-ms commented on lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb:
```diff
@@ -202,16 +201,9 @@ def self.run_release_upload(params)
-                UI.important("Could not find the excluded distribution group: #{group}") unless distribution_group_names.include? group.downcase 
-              end
+              
+              destination_array = distribution_groups.map {|h| h['name'].downcase }
```
I don't think we should downcase here (it apparently works but I wouldn't count on it)

Hanstad replied:
I agree, so I have removed it

jpeg-ms commented on lib/fastlane/plugin/appcenter/actions/appcenter_upload_action.rb:
```diff
@@ -489,12 +505,11 @@ def self.available_options
           FastlaneCore::ConfigItem.new(key: :destinations,
                                   env_name: "APPCENTER_DISTRIBUTE_DESTINATIONS",
-                               description: "Comma separated list of destination names. Both distribution groups and stores are supported. All names are required to be of the same destination type",
+                               description: "Comma separated list of destination names, use '*' for all distribution groups and exclude groups with destination_excludes. Both distribution groups and stores are supported. All names are required to be of the same destination type",
```
Oops :) Need to update comment

Hanstad replied:
Great catch!

Hanstad commented:
You are welcome and thanks for your feedback even though it was not what I wished for :)
